src/main/java/mate/academy/AsyncRequestProcessor.java

```diff
@@ -1,16 +1,35 @@
 package mate.academy;
 
+import java.util.Map;
 import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
 import java.util.concurrent.Executor;
 
 public class AsyncRequestProcessor {
+    private final Map<String, UserData> cache = new ConcurrentHashMap<>();
     private final Executor executor;
 
     public AsyncRequestProcessor(Executor executor) {
         this.executor = executor;
     }
 
     public CompletableFuture<UserData> processRequest(String userId) {
-        return null;
+        if (cache.containsKey(userId)) {
+            return CompletableFuture.completedFuture(cache.get(userId));
+        }
+        return CompletableFuture.supplyAsync(() -> {
+            sleep(500);
+            UserData userData = new UserData(userId, "Details for " + userId);
+            cache.put(userId, userData);
```

Comment:
Be aware that 'cache.put' is not an atomic operation with the preceding code. If 'processRequest' is called simultaneously with the same 'userId', there could be a race condition leading to the creation of multiple 'UserData' objects for the same 'userId'. Consider using 'cache.computeIfAbsent' to ensure atomicity.